cpp/bench/ann/src/faiss/faiss_gpu_benchmark.cu

```diff
@@ -158,5 +179,15 @@ REGISTER_ALGO_INSTANCE(std::uint8_t);
 
 #ifdef ANN_BENCH_BUILD_MAIN
 #include "../common/benchmark.hpp"
-int main(int argc, char** argv) { return raft::bench::ann::run_main(argc, argv); }
+int main(int argc, char** argv)
+{
+  rmm::mr::cuda_memory_resource cuda_mr;
+  // Construct a resource that uses a coalescing best-fit pool allocator
+  rmm::mr::pool_memory_resource<rmm::mr::cuda_memory_resource> pool_mr{&cuda_mr};
+  rmm::mr::set_current_device_resource(
```

Comment:
For single threaded benchmarks this is fine and also for multi-threaded benchmarks, the pool will be correctly shared. 

We shall consider how the RAFT handle is used in multi threaded environment. Taging @achirkin for advice.

Notes:
- FAISS keeps a raft resource handles in [StandardGpuResoures object](https://github.com/tarang-jain/faiss/blob/rmm-pool-alloc/faiss/gpu/StandardGpuResources.h#L155).
- The benchmark wrapper [has a shared pointer](https://github.com/rapidsai/raft/blob/1e4961e2354afba116e3479c5ec9041937b9922e/cpp/bench/ann/src/faiss/faiss_gpu_wrapper.h#L175) to `StandardGpuResources`.
- For multi-threaded benchmarks, each thread [gets a copy of the ANN algo](https://github.com/rapidsai/raft/blob/1e4961e2354afba116e3479c5ec9041937b9922e/cpp/bench/ann/src/common/benchmark.hpp#L294) object.

Reply:
Oh, yes, I've added the note on multithreading here https://github.com/rapidsai/raft/blob/27bf943e77ecb6daf7e38a65aff08480c3c9dcdf/cpp/bench/ann/src/faiss/faiss_gpu_wrapper.h#L171-L186

The gist is that, in the current state, we share a single faiss handle among multiple threads; this is in contrast to a new raft handle being created for every thread in raft algorithms.
We know raft handle is not thread-safe (due to stateful cublas calls). I didn't know the internals of the FAISS StandardGpuResource at the time of writing this, so I postponed this investigation to push the raft fixes. Now we know FAISS keeps the raft handle and we probably need to do something about it.

Reply:
Yes, there is just one handle for each device. https://github.com/facebookresearch/faiss/blob/5e3eae4fccb20723dbc674b3ffa005ce09afcd8d/faiss/gpu/StandardGpuResources.cpp#L432

For benchmarking purposes I have just been using a single CPU thread to prevent differences due to number of threads.

Reply:
There is a method getAlternateStreams https://github.com/facebookresearch/faiss/blob/5e3eae4fccb20723dbc674b3ffa005ce09afcd8d/faiss/gpu/StandardGpuResources.cpp#L450 that can be used to get a vector of alternate streams for the device. Perhaps we can have something similar for raft handles.

Reply:
> For benchmarking purposes I have just been using a single CPU thread to prevent differences due to number of threads.

@tarang-jain unfortunately we don't yet have automated testing for the benchmarks tool, which means we need to be vigilant about manually testing changes in the meantime. At a minimum, the tests should be run with both `search-mode=latency` and `search-mode=throughput`. The latter is going to require using separate `raft::resources`/`StandardGpuResources` for each thread.

Reply:
I think we can just create a separate `StandardGpuResources` object for each copy of `FaissGpu` wrapper. That way they will have separate raft resource handle. Since the pool allocator is set globally all the handles will share that.

Reply:
@tfeher but that would mean that the whole index would have to be copied for each thread.

Reply:
@tfeher @achirkin
I studied the benchmark wrappers of other algorithms in detail and found a fix for this: FAISS is missing two critical pieces. I have created issues: [#3424](https://github.com/facebookresearch/faiss/issues/3424) and [3425](https://github.com/facebookresearch/faiss/issues/3425) in FAISS to address these. These are easy fixes and once we have them in place, we can do the exact same thing as `raft-ann-bench`'s GGNN wrapper is currently doing: when an index is copied, change the stream of the index to the one designated for the current thread.

Reply:
Yes `setDefaultStream` will need to update RAFT handle's default stream. But you would need to make sure that the [gpu_resource_](https://github.com/rapidsai/raft/pull/2026/files#diff-0afb7367b8b5bcffa632359ec9f1cd350f5394ee280bde52ea1b86a0ea7d2220R172) is not a shared object anymore, because then everyone is trying to update the stream for the same object.

In practice we use the `GpuResource` object as a wrapper around stream. But `GpuResource` contains other resources like the Temporary memory allocator, which might grab 1.5 GiB per `GpuResource` object. That can be circumvented by setting TempMem size to 0.